test/test_dynamo_cudagraphs.py

```diff
@@ -0,0 +1,194 @@
+# Owner(s): ["module: cuda graphs"]
+
+import sys
+import torch
+from torch.testing._internal.common_utils import (
+    TestCase,
+    run_tests,
+)
+
+from unittest.mock import patch
+import functools
+
+try:
+    import torchdynamo
+    from torch.cuda._dynamo_graphs import aot_autograd_cudagraphs
+    import functorch  # noqa: F401
+
+    TEST_DYNAMO = True
+except ImportError:
+    TEST_DYNAMO = False
+
+TEST_CUDA = torch.cuda.is_available()
+
+if not TEST_CUDA or not TEST_DYNAMO:
+    print("CUDA or dynamo not available, skipping tests", file=sys.stderr)
+    TestCase = object  # noqa: F811
+
+
+def composed(*decs):
+    def deco(f):
+        for dec in reversed(decs):
+            f = dec(f)
+        return f
+
+    return deco
+
+
+def assert_aot_autograd_counter(ok=True):
+    def deco(f):
+        @functools.wraps(f)
+        def wrap(self, *args, **kwargs):
+            torchdynamo.utils.counters.clear()
+            r = f(self, *args, **kwargs)
+            c_ok = torchdynamo.utils.counters["aot_autograd"]["ok"]
+            c_not_ok = torchdynamo.utils.counters["aot_autograd"]["not_ok"]
+            if ok:
+                self.assertGreater(c_ok, 0)
+                self.assertEqual(c_not_ok, 0)
+            else:
+                self.assertEqual(c_ok, 0)
+                self.assertGreater(c_not_ok, 0)
+            return r
+
+        return wrap
+
+    return deco
+
+
+def patch_all(ok=True):
+    return composed(
+        patch("torchdynamo.config.verify_correctness", True),
+        assert_aot_autograd_counter(ok),
+    )
+
+
+N_ITERS = 5
+
+
+class TestDynamoCudaGraphs(TestCase):
+    @patch_all()
+    def test_basic(self):
+        def model(x, y):
+            return (x + y) * y
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
+            for i in range(N_ITERS):
+                x = torch.randn(3, device="cuda", requires_grad=True)
+                y = torch.randn(3, device="cuda")
+                loss = model(x, y).sum()
+                loss.backward()
+
+    @patch_all()
+    def test_dtoh(self):
+        def model(x, y):
+            a = x + y
+            b = a.cpu() * 3
+            return b
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
+            for i in range(N_ITERS):
+                x = torch.randn(3, device="cuda", requires_grad=True)
+                y = torch.randn(3, device="cuda")
+                loss = model(x, y).sum()
+                loss.backward()
+
+    @patch_all()
+    def test_htod(self):
+        def model(x, y):
+            a = x + y
+            return a * 3
+
+        with torchdynamo.optimize(aot_autograd_cudagraphs):
+            for i in range(N_ITERS):
+                x = torch.randn(3, device="cuda", requires_grad=True)
+                y = torch.randn((), device="cpu")
+                loss = model(x, y).sum()
+                loss.backward()
+
+    @patch("functorch._src.config.use_functionalize", True)
```

Comment:
noob q: I've seen the terminology functionalize a few times and my understanding is you remove stateful operations to ship to compilers that can't represent aliasing. Is that the majority of compilers? Some compilers we really care about?

Reply:
Functionalize gives you a graph that doesn't have mutating operations in it. CUDA graphs actually doesn't want functionalization, but we need it because there are passes like the partitioner we use here which are unsound in the presence of mutation.